platform/jvm/src/report_processing.rs

```diff
@@ -0,0 +1,246 @@
+  let os_brand = read_string(env, attributes, &CLIENT_ATTRS_OS_BRAND)
+    .map_err(|e| anyhow::anyhow!("failed to parse brand: {e}"))?;
+  let os_version = read_string(env, attributes, &CLIENT_ATTRS_OS_VERSION)
+    .map_err(|e| anyhow::anyhow!("failed to parse os version: {e}"))?;
+  let os_build = OSBuildArgs {
+    brand: Some(builder.create_string(&os_brand)),
+    version: Some(builder.create_string(&os_version)),
+    ..Default::default()
+  };
+  Ok(DeviceMetricsArgs {
+    manufacturer: Some(builder.create_string(&manufacturer)),
+    model: Some(builder.create_string(&model)),
+    os_build: Some(OSBuild::create(builder, &os_build)),
+    time: Some(timestamp),
+    ..Default::default()
+  })
+}
+
+fn build_app_metrics<'fbb>(
+  env: &mut JNIEnv<'_>,
+  builder: &mut FlatBufferBuilder<'fbb>,
+  attributes: &JObject<'_>,
+) -> anyhow::Result<AppMetricsArgs<'fbb>> {
+  let version_code = CLIENT_ATTRS_VERSIONCODE
+    .get()
+    .ok_or(InvariantError::Invariant)?
+    .call_method(env, attributes, ReturnType::Primitive(Primitive::Long), &[])?
+    .j()?;
+  let build_number = Some(AppBuildNumber::create(
+    builder,
+    &AppBuildNumberArgs {
+      version_code,
+      ..Default::default()
+    },
+  ));
+  let app_id = read_string(env, attributes, &CLIENT_ATTRS_APP_ID)
+    .map_err(|e| anyhow::anyhow!("failed to parse app_id: {e}"))?;
+  let app_version = read_string(env, attributes, &CLIENT_ATTRS_APP_VERSION)
+    .map_err(|e| anyhow::anyhow!("failed to parse app_version: {e}"))?;
+  Ok(AppMetricsArgs {
+    app_id: Some(builder.create_string(&app_id)),
+    version: Some(builder.create_string(&app_version)),
+    build_number,
+    ..Default::default()
+  })
+}
+
+fn read_string(
+  env: &mut JNIEnv<'_>,
+  attributes: &JObject<'_>,
+  method: &OnceLock<CachedMethod>,
+) -> anyhow::Result<String> {
+  let value = method
+    .get()
+    .ok_or(InvariantError::Invariant)?
+    .call_method(env, attributes, ReturnType::Object, &[])?
+    .l()?;
+
+  Ok(
+    unsafe { env.get_string_unchecked(&value.into())? }
+      .to_string_lossy()
+      .to_string(),
+  )
+}
+
+fn read_stream_to_file(
+  env: &mut JNIEnv<'_>,
+  stream: &JObject<'_>,
+) -> anyhow::Result<std::fs::File> {
+  let mut file = tempfile::tempfile()?;
+  let buffer = env.new_byte_array(BUFFER_SIZE)?;
+  let reader = INPUT_STREAM_READ.get().ok_or(InvariantError::Invariant)?;
+
+  loop {
+    let bytes_read = reader
+      .call_method(
+        env,
+        stream,
+        ReturnType::Primitive(Primitive::Int),
+        &[JValueWrapper::JObject(buffer.as_raw()).into()],
+      )?
+      .i()?;
+
+    if bytes_read <= 0 {
+      break;
+    }
+
+    let buffer_elements =
+      unsafe { env.get_array_elements(&buffer, jni::objects::ReleaseMode::NoCopyBack)? };
+    let file_contents = unsafe {
+      #[allow(clippy::cast_sign_loss)]
+      &*(std::ptr::from_ref::<[i8]>(&buffer_elements[.. bytes_read as usize]) as *const [u8])
```

Comment:
Hard to follow this but would it be possible to just transmute buffer_elements to &[u8]?